Fix Rationals.boundary, Rationals.is_open, Rationals.is_closed

gschintgen opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #18134

#### Brief description of what is fixed or changed
This commit fixes #18134 by
- correcting `Rationals.boundary`,
- improving the `is_open` logic by using `is_empty`
  (previously `is_open` could never return `False`),
- updating the docstrings of `is_open` and `is_closed`.

It also makes a few related tests stricter by changing
`assert not ...` to `assert ... is False`.


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* sets
  * Fixed `Rationals.boundary` to be `Reals`. (`is_open` and `is_closed` now both return `False`.)
  * `is_open` logic is improved.
<!-- END RELEASE NOTES -->

gschintgen committed:
Fix Rationals.boundary, Rationals.is_open, Rationals.is_closed

This commit fixes #18134 by
- correcting Rationals.boundary,
- improving the `is_open` logic by using `is_empty`
  (previously `is_open` could never return `False`),
- updating the docstrings of `is_open` and `is_closed`.

It also makers a few related tests stricter by changing
`assert not ...` to `assert ... is False`.

jksuom commented on sympy/sets/sets.py:
```diff
@@ -555,25 +555,29 @@ def boundary(self):
+
         A set is open if and only if it has an empty intersection with its
-        boundary.
+        boundary. The openness of a subset of the reals is determined with
```
I think that this could be more explicit. Maybe "a subset A of the reals is open if and only if each one of its points is contained in an open interval that is a subset of A". (This is equivalent to saying that A is the union of a family of open intervals.)

gschintgen replied:
Thanks for the suggestion, I've changed it.
I think I'll also rephrase the `is_closed` docstring to:
"A set is closed if its complement is an open set. In a metric space, in particular the reals, a set is closed if and only if it contains all of its limit points."

And I'll also change the added union test to test the actual results instead of only the fact that they're simplified.

gschintgen committed:
Improve unions of number sets (Rationals + Reals, etc.)

Before this commit some trivial number set unions were not simplified:
Naturals0 + Naturals
Rationals + Naturals
Rationals + Naturals0
Reals + Naturals
Reals + Naturals0
Reals + Rationals

This is addressed by adding handlers in sets/handlers/union.py.

(Plus: rewording of is_open docstring and two more tests)

jksuom commented:
Looks good.
